sycl/include/CL/sycl/detail/pi.hpp

```diff
@@ -129,56 +131,79 @@ inline void piCheckResult(PiResult pi_result) {
   piCheckThrow<cl::sycl::runtime_error>(pi_result);
 }
 
-template <typename FnType> class Trace {
+#define PI_TRACE_ENABLED (std::getenv("SYCL_PI_TRACE") != nullptr)
+
+template <typename FnType, size_t FnOffset> class Trace {
 private:
   FnType m_FnPtr;
-  static bool m_TraceEnabled;
+  std::string m_FnName;
 
 public:
-  Trace(FnType FnPtr, const std::string &FnName) : m_FnPtr(FnPtr) {
-    if (m_TraceEnabled)
-      std::cout << "---> " << FnName << "(";
-  }
-
+  Trace();
   template <typename... Args> PiResult operator()(Args... args) {
-    if (m_TraceEnabled)
+    if (PI_TRACE_ENABLED) {
+      std::cout << "---> " << m_FnName << "(";
       printArgs(args...);
+    }
 
-    auto r = m_FnPtr(args...);
+    PiResult r = m_FnPtr(args...);
 
-    if (m_TraceEnabled) {
+    if (PI_TRACE_ENABLED) {
       std::cout << ") ---> ";
       std::cout << (print(r), "") << std::endl;
     }
     return r;
   }
 };
 
-template <typename FnType>
-bool Trace<FnType>::m_TraceEnabled = (std::getenv("SYCL_PI_TRACE") != nullptr);
+template <typename FnType, size_t FnOffset>
+class TraceCheck : private Trace<FnType, FnOffset> {
+public:
+  TraceCheck() : Trace<FnType, FnOffset>(){};
+
+  template <typename Exception = cl::sycl::runtime_error, typename... Args>
+  void operator()(Args... args) {
+    PiResult Err = (Trace<FnType, FnOffset>::operator()(args...));
+    piCheckThrow<Exception>(Err);
+  }
+};
+
+// Explicit specialization declarations for Trace class for every FnType.
+// The offsetof is used as a template argument to uniquely identify every
+// api.
+#define _PI_API(api)                                                           \
+  template <>                                                                  \
+  Trace<decltype(&::api),                                                      \
+        (offsetof(_pi_plugin::FunctionPointers, api))>::Trace();
```

Comment:
Is the implementation missing here (might be further below)?

Reply:
@bjoernknafla Can you explain your suggestion again?